INT-4005: Do `capacity` attr invalid (persistence)

manueljordan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4005
- The `<int:queue>` lets declare the `capacity` attribute together
  with either `message-store` or `ref` attributes, which it is not
  correct

manueljordan committed:
INT-4005: Do `capacity` attr invalid (persistence)

JIRA: https://jira.spring.io/browse/INT-4005

* The `<int:queue>` lets declare the `capacity` attribute together
with either `message-store` or `ref` attributes, which it is not
correct

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/PointToPointChannelParser.java:
```diff
@@ -55,14 +55,19 @@ protected BeanDefinitionBuilder buildBeanDefinition(Element element, ParserConte
-			}
+				if (hasCapacity && hasStoreRef) {
+					parserContext.getReaderContext().error(
+							"The 'capacity' attribute is not allowed" + " when providing a 'message-store' to a custom MessageGroupStore.",
```
Minor. Just for future PRs to save some time: https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-Framework-Code-Style
We prefer line length no more than 120 symbols.

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidChannelWithCapacityAndQueueParserTests-context.xml:
```diff
@@ -0,0 +1,13 @@
+			http://www.springframework.org/schema/integration/spring-integration.xsd">
+
+	<channel id="output">
+		<queue ref="queue" capacity="10" />
```
I think for the consistency we have to declare the `queue` bean:
````xml
<beans:bean id="queue" class="java.util.concurrent.LinkedBlockingQueue" />
````
Otherwise this config looks like invalid by the absent bean reason.

manueljordan replied:
Agree, I did realize that for my new files an error message had appear, but not in yours (for `InvalidChannelWithMessageStoreParserTests-context`)

artembilan replied:
No reason to comment if you agree with me.
That saves some our time, too :smile:

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidChannelWithCapacityAndQueueParserTests.java:
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2002-2015 the original author or authors.
```
This is a new class, so only the current year.

manueljordan commented:
Oh... understood... sorry, I did not realise about that. 

If this PR is approved, I must create a new JIRA, for `<int:priority-queue  message-store="jdbcPriorityChannelMessageStore" />` where `capacity` is allowed with `message-store` too.

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidChannelWithCapacityAndQueueParserTests.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @author Manuel Jordan
+ */
+public class InvalidChannelWithCapacityAndQueueParserTests {
```
Since we are already far away from the beginning of the project, it is a rule to include the current version `@since 4.3`

manueljordan replied:
Ok Understood

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidChannelWithCapacityAndQueueParserTests.java:
```diff
@@ -0,0 +1,42 @@
+	@Test
+	public void testQueueCapacityAndRefIllegal() throws Exception {
+		exception.expect(BeanDefinitionParsingException.class);
+		exception.expectMessage(Matchers.containsString("'capacity' attribute is not allowed"));
```
Even if it is test class, I prefer to add `this.` just to work out the good habit, when I come to the `main` classes.

manueljordan replied:
I understand, I did use how reference the code structure from `InvalidChannelWithMessageStoreParserTests`

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidChannelWithCapacityAndMessageStoreParserTests.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @author Manuel Jordan
+ */
+public class InvalidChannelWithCapacityAndMessageStoreParserTests {
```
Actually, I'd move these new classes to the existing `InvalidChannelWithMessageStoreParserTests` as additional test methods. Just because all of them are black boxes to each other, but have similar premise.
Right, I might rename it to reflect the new reality, but anyway it looks like all these three test cases can live in the same logic place.

manueljordan replied:
Totally agree, I thought the same, but, since I couldn't find an example where I can be able to indicate what channel (according with the id) should be test for each situation, I have created these new test classes.

I mean the .xml file should have 2 or 3 channels, each one representing an invalid scenario, therefore, each `Test` method should retrieve each channel by id.

artembilan replied:
No, that won't work that way. 
Entire config will fail on the first invalid case.
You still have to have 3 configs, and each test method will load its own config.
That's why I said that all of them are black boxes to each other and they don't share one of them like you can find in the many valid tests.

manueljordan replied:
Oh Ok... so we must create a test class with its respective xml file to test each scenario. Perhaps verbose in some way, but this approach keeps each scenario isolated.

artembilan replied:
No, a bit different way.
One existing `InvalidChannelWithMessageStoreParserTests` with 2 new test-methods for your new `InvalidChannelWithCapacityAndMessageStoreParserTests-context.xml` and `InvalidChannelWithCapacityAndQueueParserTests-context.xml`. And they all even will share that `ExpectedException` `@Rule`!

manueljordan replied:
Just curious if already exists a test class with that approach, to take it how example. Normally I use a test class and import just one Configuration class or one xml file for this kind of scenario.

artembilan replied:
Look. We are going to test that our attributes restriction works well showing an error on parsing the XML. Therefore it fully fails the entire application context.
That's why the code in the test-method looks like:
```
new ClassPathXmlApplicationContext(getClass().getSimpleName() + "-context.xml", getClass()).close();
```
But having the `@Rule ExpectedException` we definitely try to be sure that the `@Test` will pass with desired error.
Those new `@Test` to the same class won't break anything, each test-method is isolated from any other with its own attempt to load its own context.
For example `ChainParserTests#invalidNestedChainWithLoggingChannelAdapter()`.

manueljordan replied:
> For example ChainParserTests#invalidNestedChainWithLoggingChannelAdapter().

Thanks, I've looked the code, tomorrow I will start try to do the experiment. I will try to test the three invalid scenarios. 

One thing I've noticed, why the `src/test/resources` is not used to put there the .xml files for testing?

artembilan replied:
For convenience purpose. Typically we have a pair - test class and config for it with the same name and suffix `-context.xml`. So, in the same directory they are alongside. And it is enough easy to jump to them from the IDE.
Using `src/test/resources` we introduce some inconvenience for jumping between them.

manueljordan replied:
Thanks by the explanation, understood..

artembilan commented:
Pay attention to the red report from Travis CI. 
It is really helpful tool, especially when we switched on [Checkstyle](http://checkstyle.sourceforge.net/) :smile: :
```
:spring-integration-core:checkstyleMain[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/config/xml/PointToPointChannelParser.java:70: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
```
The `Trailing whitespace` can be fixed as an IDE configuration: 
https://github.com/spring-projects/spring-framework/tree/master/src/eclipse
https://github.com/spring-projects/spring-framework/tree/master/src/idea

All concerns are minor and I can address them on merge.
Although you feel free to do them yourself just for the homework exercise.
I can wait.

But I'm sure you have enough other works and you'll do the next PR much better.
Although already this one is pretty good. And I even know that other guys may merge it as is.
My idea for the good code style and design nit-picking is to avoid such a discussions in the future.

Thank you for the contribution!
Looking forward for more!

artembilan commented:
> . sorry, I did not realise about that. 

No any apologies! You're doing fine. :kimono: 

> I must create a new JIRA, for <int:priority-queue message-store="jdbcPriorityChannelMessageStore" /> where capacity is allowed with message-store too.

Good catch! You can address that problem directly in this PR. The issue is orthogonal the same and the place to fix is still exactly the same `PointToPointChannelParser`.

Pay attention: I didn't noticed that even myself :wink:

artembilan commented:
BTW, sorry for so much new info around.

Some guys are really annoyed for all these our Open Source process rules... :smile:

manueljordan commented:
Hi Artem

Thanks by your support and understanding (learning in real action)

>Pay attention to the red report from Travis CI. 
Ok, I have seen the link, it is the unique error. about the `Trailing whitespace`

> Although you feel free to do them yourself just for the homework exercise.
> I can wait.
Let me do this, I must learn.

I am confused about the Eclipse link, it refers about `Change Eclipse Settings to allow forbidden references` and the problem is about `Trailing whitespace`, the link is correct?

Sorry by my innocence, I am with the impression to replace that files in eclipse (my STS installation) If I am correct, where? I can't find that files.

I need your guidance.

manueljordan commented:
> No any apologies! You're doing fine.

Thanks :)


> Good catch! You can address that problem directly in this PR. 

Understood.

> BTW, sorry for so much new info around.
> Some guys are really annoyed for all these our Open Source process rules... :smile:

In some way perhaps, But I am fine.... thanks by your patience 

These rules exists by an important reason, keep an excellent standard in Spring source code.

artembilan commented:
I so sorry for misleading you about that `Trailing whitespace`.
I just meant there that we have some configs for IDE there, which include the `Trailing whitespace` option, too.

Not sure how to do that in the STS, but in my IDEA I only import that file from the particular setting.

manueljordan commented:
Hi, about the `Trailing whitespace`, I have configured
[How to auto-remove trailing whitespace in Eclipse?](http://stackoverflow.com/questions/1043433/how-to-auto-remove-trailing-whitespace-in-eclipse). Tomorrow, lets see what happens.

manueljordan commented:
Tomorrow to fix all this: 

I am assuming that after to do the new editions, I should create a new commit in my local repository, push my branch to my remote repository and again re send a pull request. 

If I am wrong, what is the expected approach?

artembilan commented:
The correct approach is just:
- make local changes and push to your remote.

That's all. GitHub picks up new commit in the remote branch and reflects them immediately in the PR.

manueljordan commented:
Interesting 

> make local changes and push to your remote.

Seems missing do a local commit before to do a push, but according with

> That's all. GitHub picks up new commit in the remote branch and reflects them immediately in the PR.

Something internal in github happens...  interesting

manueljordan committed:
INT-4005: Do `capacity` att invalid (persistence)

JIRA: https://jira.spring.io/browse/INT-4005

* The `<int:queue>` lets declare the `capacity` attribute together
with either `message-store` or `ref` attributes, which it is not
correct
* The `priority-queue` lets declare the `capacity` attribute
together with the `message-store` attribute, which it is not
correct
* Reference documentation fixed for `channel.adoc` and `jdbc.adoc`,
doing mention about these restrictions

artembilan commented:
@manueljordan , please, look into #1794 regarding the issue in the HTTP module we have noticed today.
I think @garyrussell , will review it tomorrow and merge.
You will be forced to `rebase` your branch to the latest `master` after that.

manueljordan committed:
INT-4005: Do `capacity` attr invalid - persistence

JIRA: https://jira.spring.io/browse/INT-4005

* Some methods from the `DefaultHttpHeaderMapperFromMessageOutboundTests`
class has been `@Ignored` due the `INT-4017` situation

JIRA: https://jira.spring.io/browse/INT-4017

manueljordan commented:
@artembilan I did the push without read your previous comment. Perhaps should be merge from my commit first.... anyway I wait for instructions.

garyrussell commented:
The HTTP fix is merged

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/PointToPointChannelParser.java:
```diff
@@ -30,6 +28,7 @@
 import org.springframework.integration.store.MessageGroupQueue;
 import org.springframework.util.StringUtils;
 import org.springframework.util.xml.DomUtils;
+import org.w3c.dom.Element;
```
Hello, @manueljordan !

Thank you for update!

As you already noticed Gary's comment about HTTP tests, you should `rebase` to the latest master and "un"`@Ignore` those tests in the `DefaultHttpHeaderMapperFromMessageOutboundTests`.

But that's not all. I'll let you know what to change else during this review phase.

So, this one imports order change is wrong. See https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-Framework-Code-Style#import-statements. 
Right, it is minor and I can fix on merge. But to follow with our style helps minimize the code for review and save some time for all of us. At least I'll stop to write similar comments here :smile:

manueljordan replied:
Hi @artembilan 

>So, this one imports order change is wrong

I see, but just I did copy from other classes... Seems the problem comes from
[Removing whitespace...](http://stackoverflow.com/a/2618521)

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidChannelWithMessageStoreParserTests-context.xml:
```diff
@@ -10,6 +10,7 @@
 	</channel>
 
+	<beans:bean id="queue" class="java.util.concurrent.LinkedBlockingQueue" />
 	<beans:bean id="messageStore" class="org.springframework.integration.store.SimpleMessageStore" />
```
Minor, but would be better do not stick together logically separate beans

manueljordan replied:
Ok, add an empty line...

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidChannelWithMessageStoreParserTests.java:
```diff
@@ -17,22 +17,26 @@
+ *             InvalidPriorityChannelParserTests
  */
+@Deprecated
 public class InvalidChannelWithMessageStoreParserTests {
```
No, it is an overhead. 
This is just test and it isn't a public API, so if you decided to change its name, just do that!
And I understand your point here and know that it was my advice to collect similar tests together.
LGTM for that part!

If we don't need this test at all just remove.
Since you copy/paste an existing tests to a new one class, pleas, show respect and leave Dave Syer's name there as well :wink:

manueljordan replied:
>If we don't need this test at all just remove.
>Since you copy/paste an existing tests to a new one class, 
>pleas, show respect and leave Dave Syer's name there as well

That's why I did use that annotation, ok I will remove that test class and add the author note in the new version

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidPriorityChannelParserTests.java:
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2002-2016 the original author or authors.
```
A new class just with current year.

manueljordan replied:
I forget that.... understood.

artembilan replied:
I said you before: don't comment if you agree. 
Just fix and add simple commend `Pushed`.

Only different opinion makes sense. :smile:

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidPriorityChannelParserTests.java:
```diff
@@ -0,0 +1,51 @@
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.springframework.beans.factory.parsing.BeanDefinitionParsingException;
```
Something is still wrong with import settings.
Need to be a blank line in between `org.*` and `org.springframework.*`

manueljordan replied:
I have confirmed that, it is because [Removing whitespace...](http://stackoverflow.com/a/2618521). I must open the Atom editor and add the blank lines... it complicates the work as you can see... I don't know what extra configuration I must do.

artembilan replied:
M-m-m. I don't know how STS works, but when I run reformat code/imports in the IntelliJ IDEA, it does that exactly how I have configured it before.
So, if there was no blank line in between those imports, IDEA inserts them.
I hope STS should do something similar.

At least @garyrussell uses STS and we working together for a long time already without so much issues around code formatting.

manueljordan replied:
I can see in real time the required blank line between the packages, but when I save the file, Eclipse removes the blank line, so I must use atom editor, so how you can see, that file was not fixed manually by me. 

Other problem is when I do a copy paste of packages from one class to other Eclipse re-arrange the import package declarations. ....

garyrussell replied:
Here is my STS import config - it automatically adds a blank line between groups:

![screen shot 2016-04-29 at 11 49 28 am](https://cloud.githubusercontent.com/assets/483832/14921512/a16b9c42-0e00-11e6-8ad1-f2112d33dee8.png)

garyrussell replied:
And java "save actions"...

![screen shot 2016-04-29 at 11 52 03 am](https://cloud.githubusercontent.com/assets/483832/14921565/db6da20a-0e00-11e6-9f14-0e5673b94d55.png)

artembilan replied:
Thank you Gary!
That's much better to demonstrate it via screenshots, than try to explain via words.
WDYT if we add them to the https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-Framework-Code-Style, if that may be such a challenge for new contributors?

manueljordan replied:
Thanks a lot @garyrussell 
For the image shown above, about the `Additional actions`, could you press the `Configure` button and show me an image for `Code Organizing` tab? Thanks!

manueljordan replied:
Totally agree about the @artembilan suggestion. How we can see I am with this problem due Eclipse default configuration. I think that note should be updated to all the Spring project too...

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidPriorityChannelParserTests.java:
```diff
@@ -0,0 +1,51 @@
+		new ClassPathXmlApplicationContext("InvalidPriorityChannelWithComparatorAndMessageStoreParserTests.xml",
+				getClass()).close();
+	}
+}
```
I'd prefer to have a blank line just before the last `}` in each class.
That differentiate methods visually

manueljordan replied:
understood

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/InvalidPriorityChannelWithComparatorAndMessageStoreParserTests.xml:
```diff
@@ -0,0 +1,16 @@
+	</channel>
+
+	<beans:bean id="messageStore" class="org.springframework.integration.store.SimpleMessageStore" />
+	<beans:bean id="comparator" class="org.springframework.integration.aggregator.MessageSequenceComparator" />
```
DITO for blank line in between beans.

manueljordan replied:
Understood

artembilan commented on src/reference/asciidoc/jdbc.adoc:
```diff
@@ -433,6 +433,8 @@ For this purpose, the `INT_CHANNEL_MESSAGE` has a `MESSAGE_PRIORITY` column to s
 In addition, a new `MESSAGE_SEQUENCE` column is also provided to achieve a robust first-in-first-out (FIFO) polling mechanism, even when multiple messages are stored with the same priority in the same millisecond.
 Messages are polled (selected) from the database with `order by MESSAGE_PRIORITY DESC NULLS LAST, CREATED_DATE, MESSAGE_SEQUENCE`.
 
+IMPORTANT: The `capacity` attribute is not allowed when the `message-store` attribute is used.
```
??? I think you place this sentence in the JDBC chapter accidentally.
I'm sure you meant the 
```
 [[channel-configuration-prioritychannel]]
 ===== PriorityChannel Configuration
```
section in the `channel.adoc`

manueljordan replied:
I think is correct: just above of that paragraph is located the `Priority Channel` section, this `capacity` restriction applies for PriorityChannel as well. 

Furthermore, just below there is an example showing QueueChannel and PriorityChannel working with together with `JdbcChannelMessageStore` (using `message-store`). So that note applies for both queue and priority.

artembilan replied:
Yes. I understand, but this your new sentence does not make sense in context of JDBC.
More over it looks like out-of-context when I read the whole text about `JdbcChannelMessageStore`. 
That's why I don't see reason in this sentence here. But I won't mind if it will be present in the `channel.adoc` for `PriorityChannel`.

Please, elaborate more if I am missing anything else.
Thanks

manueljordan replied:
Ok, removed that line from `jdbc.adoc`. 

About 

```
 [[channel-configuration-prioritychannel]]
 ===== PriorityChannel Configuration
```

I can't put the line 

```
IMPORTANT: The `capacity` attribute is not allowed when the `message-store` attribute is used.
``` 

it because that section for Priority Channel does not cover nothing about `message-store`, so that is confuse, but just below in the following paragraph I have added (already pushed)

```
Since _version 4.0_, the `priority-channel` child element supports the `message-store` option (`comparator` and `capacity` are not allowed in that case).
```

Practically:
```
(`comparator` and `capacity` are not allowed in that case)
```

artembilan replied:
OK. That's fine. Please, just remove this sentence here.
Anyway I see there the `message-store` section:
>  The message store must be a `PriorityCapableChannelMessageStore` and, in this case, the namespace parser will declare a `QueueChannel` instead of a `PriorityChannel`.

OK. Never mind. Just push what you find as OK and we will be very close to merge and go ahead with other stuff!

artembilan commented:
Well, @manueljordan , actually that's all.
I can fix all those minor concerns on merge, but you should let me know first of all.
We won't mind if you would like still to play with that.
There is no hurry: we are only in the beginning of current deadline :smile:

manueljordan commented:
Let me fix that please, my unique problem is the Eclipse/STS problem about the `trailing whitespace`, now does problem about the package declaration....

manueljordan committed:
INT-4005: Do `capacity` attr invalid - persistence

JIRA: https://jira.spring.io/browse/INT-4005

* Minor changes about conventions and formats applied.

manueljordan committed:
INT-4005: Do `capacity` attr invalid - persistence

JIRA: https://jira.spring.io/browse/INT-4005

* Removed the `@Ignore` for the
`DefaultHttpHeaderMapperFromMessageOutboundTests` class

artembilan commented:
Yep! Now it's ready for merge.
My congratulations for the good hard work :smile: 

Looking forward for more!

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/PointToPointChannelParser.java:
```diff
@@ -55,13 +56,19 @@ protected BeanDefinitionBuilder buildBeanDefinition(Element element, ParserConte
+				if (hasCapacity && hasStoreRef) {
+					parserContext.getReaderContext().error(
+							"The 'capacity' attribute is not allowed"
+									+ " when providing a 'message-store' to a custom MessageGroupStore.",
```
Minor: the `+` must be on the previous line.

manueljordan replied:
Let me fix it

artembilan replied:
It's late! I'm on it already during merge with some other simple code style polishing around.
BTW, you have not to do any code style refactoring during your PR. That will simplify our review life :smile:

manueljordan commented:
Thanks a lot!... :)

manueljordan committed:
INT-4005: Do `capacity` attr invalid - persistence

JIRA: https://jira.spring.io/browse/INT-4005

* Minor change about conventions and formats applied

artembilan commented:
You still have some problems with whitespaces:
> :spring-integration-core:checkstyleMain[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/config/xml/PointToPointChannelParser.java:68: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]

Plus one more note for the future: please reflect in each commit exactly its changes. 
The first commit might become not relevant to subsequent commits.

manueljordan commented:
Let me fix it...

artembilan commented:
No, that's late! I said you already.
Merging...
Relax! :smile:

manueljordan commented:
Oh ok.... I did the instructions on Eclipse(STS) about the configurations shared on the images... wondered why it happens again...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/8cfb7414ada0ed84112dd34f4eb00960a937c241
